src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,612 @@
+            int mapped_schema_index = found->s_id;
+            schema_map[mapped_schema_index] = schema_index;
+            num_children_map[mapped_schema_index] = 0;
+          }
+        }
+
+        if (num_children > 0) {
+          tree_stack.push_back(found);
+          num_children_stack.push_back(num_children);
+        } else {
+          // go back up the stack/tree removing children until we hit one with more children
+          bool done = false;
+          while (!done) {
+            int parent_children_left = num_children_stack.back() - 1;
+            if (parent_children_left > 0) {
+              num_children_stack.back() = parent_children_left;
+              done = true;
+            } else {
+              tree_stack.pop_back();
+              num_children_stack.pop_back();
+            }
+ 
+            if (tree_stack.size() == 0) {
+              done = true;
+            }
+          }
+        }
+      }
+
+      // If there is a column that is missing from this file we need to compress the gather maps
+      //  so there are no gaps
+      std::vector<int> final_schema_map;
+      final_schema_map.reserve(schema_map.size());
+      for (auto it = schema_map.begin(); it != schema_map.end(); it++) {
+        final_schema_map.push_back(it->second);
+      }
+
+      std::vector<int> final_num_children_map;
+      final_num_children_map.reserve(num_children_map.size());
+      for (auto it = num_children_map.begin(); it != num_children_map.end(); it++) {
+        final_num_children_map.push_back(it->second);
+      }
+
+      std::vector<int> final_chunk_map;
+      final_chunk_map.reserve(chunk_map.size());
+      for (auto it = chunk_map.begin(); it != chunk_map.end(); it++) {
+        final_chunk_map.push_back(it->second);
+      }
+
+      return column_pruning_maps{final_schema_map, final_num_children_map, final_chunk_map};
+    }
+
+private:
+
+    void add_depth_first(const std::vector<std::string> & names, const std::vector<int> & num_children, int parent_num_children) {
+      CUDF_FUNC_RANGE();
+      if (parent_num_children == 0) {
+        // There is no point in doing more the tree is empty, and it lets us avoid some corner cases
+        // in the code below
+        return;
+      }
+      int local_s_id = 0; // There is always a root on the schema
+      int local_c_id = -1; // for columns it is just the leaf nodes
+      auto num = names.size();
+      std::vector<column_pruner*> tree_stack;
+      std::vector<int> num_children_stack;
+      tree_stack.push_back(this);
+      num_children_stack.push_back(parent_num_children);
+      for(uint64_t i = 0; i < num; i++) {
+        auto name = names[i];
+        auto num_c = num_children[i];
+        local_s_id++;
+        int tmp_c_id = -1;
+        if (num_c == 0) {
+          // leaf node...
+          local_c_id++;
+          tmp_c_id = local_c_id;
+        }
+        tree_stack.back()->children.try_emplace(name, local_s_id, tmp_c_id);
```

Comment:
curious, why `try_emplace`, instead of `emplace`?

Reply:
`try_emplace` does not replace the value if it is already there. Technically there is nothing stopping someone from having a malformed schema with multiple entries that have the same name. If that happens the duplicate names are combined together instead of last one wins and stops over the top of it. I could instead throw an exception if I see duplicates. Probably should add that in as some of the follow on work.